knowledge_runtime/knowledge_runtime/services/duckdb_generator.py

```
@@ -0,0 +1,711 @@
+        if not sheet_names:
+            raise RuntimeError("No sheets found in the Excel file")
+
+        base_name = self._extract_table_name(source_file)
+        safe_path = str(source_path).replace("'", "\\'")
+
+        if len(sheet_names) == 1:
+            # Single sheet: use filename as table name
+            table_name = self._sanitize_table_name(base_name)
+            conn.execute(
+                f'CREATE TABLE "{table_name}" AS '
+                f"SELECT * FROM read_xlsx('{safe_path}', header=true)"
```

Ruff (0.15.12)

[error] Possible SQL injection vector through string-based query construction

(S608)